cpp/src/Slice/MetadataValidation.h

```diff
@@ -0,0 +1,101 @@
+        Definitions,
+
+        /// The metadata can be applied to definitions of types and to where those types are referenced.
+        /// Ex: both of these are valid: `["java:buffer"] sequence<string> S;` and `["java:buffer"] StringSeq op();`.
+        ///
+        /// Note that for metadata in this category, if it has been applied to an operation, we validate that metadata
+        /// for the operation's **return type**, instead of for the operation itself.
+        /// We do the same thing for metadata applied to parameters and data members as well.
+        /// Due to the syntax of Slice, it's ambiguous whether metadata is applied to these elements or their types.
+        DefinitionsAndTypeReferences,
+
+        /// The metadata can only be applied to the types of operations parameters (including return types).
+        /// Ex: `void op(["cpp:array"] StringSeq s);` is valid, but `["cpp:array"] sequence<string> S;` is not.
+        ///
+        /// Note that for metadata in this category, if it has been applied to an operation, we validate that metadata
+        /// for the operation's **return type**, instead of for the operation itself.
+        /// We do the same thing for metadata applied to parameters and data members as well.
+        /// Due to the syntax of Slice, it's ambiguous whether metadata is applied to these elements or their types.
+        ParameterTypeReferences,
+    };
+
+    /// Alias for functions which can provide additional validation to certain metadata directives.
+    ///
+    /// They take a reference to the instance of metadata we're validating, and a reference to what it was applied on.
+    /// They should return `nullopt` to signal that the metadata was valid. If the metadata is invalid however, they
+    /// should return a string describing the error. This string will be supplied to the user as part of a diagnostic.
+    using ValidationFunc = std::function<std::optional<std::string>(const MetadataPtr&, const SyntaxTreeBasePtr&)>;
+
+    struct MetadataInfo
+    {
+        /// A list of types that this metadata can validly be applied to.
+        ///
+        /// If this list is empty we don't perform any automatic checking of whether this metadata is validly applied.
+        /// Usually, this is because determining validity isn't as straightforward as matching against a list,
+        /// and requires a more complex approach, which is achieved through providing an `extraValidation` function.
+        std::list<std::reference_wrapper<const std::type_info>> validOn;
+
+        /// Specifies how many, and what kinds of arguments, this metadata accepts.
+        MetadataArgumentKind acceptedArguments;
+
+        /// This field stores the specific values that can be provided as arguments to this metadata.
+        /// If this field is unset, then we perform no validation of the arguments (i.e. arguments can have any value).
+        std::optional<StringList> validArgumentValues = std::nullopt;
```

Comment:
Clarify that validArgumentValues has an effect only when acceptedArguments is SingleArgument or AnyNumberOfArguments.

Reply:
Fixed.